lib/src/test/resources/http4s/form-params-018.txt

```diff
@@ -0,0 +1,114 @@
+    org.http4s.QueryParamDecoder.fromUnsafeCast[java.time.Instant](p => java.time.Instant.parse(p.value))("java.time.Instant")
+
+  implicit lazy val localDateQueryParamDecoder: org.http4s.QueryParamDecoder[java.time.LocalDate] =
+    org.http4s.QueryParamDecoder.fromUnsafeCast[java.time.LocalDate](p => java.time.LocalDate.parse(p.value))("java.time.LocalDate")
+
+  implicit lazy val uuidQueryParamDecoder: org.http4s.QueryParamDecoder[java.util.UUID] =
+    org.http4s.QueryParamDecoder.fromUnsafeCast[java.util.UUID](p => java.util.UUID.fromString(p.value))("java.util.UUID")
+
+
+  object ApiVersion {
+    val ApiVersionMajor = {
+      import org.http4s.syntax.string._
+      "X-Apidoc-Version-Major".ci
+    }
+
+    def apply(req: org.http4s.Message[cats.effect.IO]): Boolean = req.headers.get(ApiVersionMajor) match {
+      case Some(v) if v.value == "0" => true
+      case _ => false
+    }
+  }
+
+
+}
+
+trait ModelRoutes {
+  import Matchers._
+
+  implicit def circeJsonDecoder[A](implicit decoder: _root_.io.circe.Decoder[A]) = org.http4s.circe.jsonOf[cats.effect.IO, A]
+  implicit def circeJsonEncoder[A](implicit encoder: _root_.io.circe.Encoder[A]) = org.http4s.circe.jsonEncoderOf[cats.effect.IO, A]
+
+  case class PostRequest(id: String, model: io.apibuilder.http4s.test.models.Model, enum: io.apibuilder.http4s.test.models.Enum, opt: _root_.scala.Option[Int], list: Seq[String], map: Map[String, String])
+
+  implicit val PostRequestDecoder: _root_.io.circe.Decoder[PostRequest] = _root_.io.circe.Decoder.instance { a =>
+    for {
+      id <- a.downField("id").as[String]
+      model <- a.downField("model").as[io.apibuilder.http4s.test.models.Model]
+      enum <- a.downField("enum").as[io.apibuilder.http4s.test.models.Enum]
+      opt <- a.downField("opt").as[_root_.scala.Option[Int]]
+      list <- a.downField("list").as[Seq[String]]
+      map <- a.downField("map").as[Map[String, String]]
+    } yield {
+      PostRequest(
+        id = id,
+        model = model,
+        enum = enum,
+        opt = opt,
+        list = list,
+        map = map
+      )
+    }
+  }
+
+  sealed trait PostResponse
+
+  object PostResponse {
+    case class HTTP200(headers: Seq[org.http4s.Header] = Nil) extends PostResponse
+  }
+
+  def post(
+    _req: org.http4s.Request[cats.effect.IO],
+    id: String,
+    model: io.apibuilder.http4s.test.models.Model,
+    enum: io.apibuilder.http4s.test.models.Enum,
+    opt: _root_.scala.Option[Int],
+    list: Seq[String],
+    map: Map[String, String]
+  ): cats.effect.IO[PostResponse]
+
+  def apiVersionMatch(req: org.http4s.Message[cats.effect.IO]): Boolean = ApiVersion(req)
+
+  def service() = org.http4s.HttpService[cats.effect.IO] {
+    case _req @ POST -> Root / "test" if apiVersionMatch(_req) =>
+    if (_req.contentType.exists(_.mediaType == _root_.org.http4s.MediaType.`application/json`)) {
+      _req.as[_root_.io.circe.Json].flatMap{
+        _.as[PostRequest].map {
```

Comment:
It should be possible to combine the parsing and type conversion via `_req.attemptAs[PostRequest].value` which returns a `Task[Either[DecodeFailure, PostRequest]]`